docs/topics/multiplatform/multiplatform-full-stack-app.md

```diff
@@ -0,0 +1,991 @@
+
+    ![New shopping list rendering](new-shopping-list-rendering.png){width=500}
+
+#### Add an input field component
+
+Next, allow users to add new entries to the shopping list using a text input field. You'll need an input component that
+provides a callback when users submit their entry to the shopping list to receive input.
+
+1. Create the `src/jsMain/kotlin/InputComponent.kt` file and fill it with the following definition:
+
+    ```kotlin
+    import org.w3c.dom.HTMLFormElement
+    import react.*
+    import org.w3c.dom.HTMLInputElement
+    import react.dom.events.ChangeEventHandler
+    import react.dom.events.FormEventHandler
+    import react.dom.html.InputType
+    import react.dom.html.ReactHTML.form
+    import react.dom.html.ReactHTML.input
+    
+    external interface InputProps : Props {
+        var onSubmit: (String) -> Unit
+    }
+    
+    val inputComponent = FC<InputProps> { props ->
+    val (text, setText) = useState("")
+    
+        val submitHandler: FormEventHandler<HTMLFormElement> = {
+            it.preventDefault()
+            setText("")
+            props.onSubmit(text)
+        }
+    
+        val changeHandler: ChangeEventHandler<HTMLInputElement> = {
+            setText(it.target.value)
+        }
+    
+        form {
+            onSubmit = submitHandler
+            input {
+                type = InputType.text
+                onChange = changeHandler
+                value = text
+            }
+        }
+    }
+    ```
+
+    The `inputComponent` keeps track of its internal state (what the user has typed so far) and exposes an `onSubmit`
+    handler that gets called when the user submits the form (usually by pressing the `Enter` key).
+
+2. To use this `inputComponent` from the application, add the following snippet to `src/jsMain/kotlin/App.kt` at the
+   bottom of the `FC` block (after the closing brace for the `ul` element):
+
+    ```kotlin
+    inputComponent {
+        onSubmit = { input ->
+            val cartItem = ShoppingListItem(input.replace("!", ""), input.count { it == '!' })
+            scope.launch {
+                addShoppingListItem(cartItem)
+                shoppingList = getShoppingList()
+            }
+        }
+    }
+    ```
+
+   * When users submit text, a new `ShoppingListItem` is created. Its priority is set to be the number of exclamation
+     points in the input, and its description is the input with all exclamation points removed. This turns `Peaches!! 🍑`
+     into a `ShoppingListItem(desc="Peaches 🍑", priority=2)`.
+   * The generated `ShoppingListItem` gets sent to the server with the client you've built before.
+   * Then, the UI is updated by obtaining the new list of `ShoppingListItem`s from the server, updating the application
+     state, and letting React to re-render the contents.
+
+#### Implement item removal
+
+Add the ability to remove items from the list so that it doesn't get too long. You can modify an existing list rather
```

Comment:
Since it's a todo list, probably also "to indicate done items / to remove items that are done"